### PR TITLE
Add addFriend endpoint

### PR DESCRIPTION
**Pull request by olegLevochkin**

### CODE REVIEW

**Review by Vlad6501 — APPROVED** [review]

File: src/main/java/com/kpi/project/model/User.java [hunk 1]
```diff
@@ -54,6 +54,10 @@
     @Enumerated(EnumType.STRING)
     Set<Role> roles;
 
+    @ElementCollection(fetch = FetchType.EAGER)
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/model/User.java:57]: Maybe Lazy

File: src/main/java/com/kpi/project/model/dto/UserDto.java [hunk 2]
```diff
@@ -21,4 +21,6 @@
 
     Set<String> roles;
 
+    String friend;
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/model/dto/UserDto.java:24]: friendName

File: src/main/java/com/kpi/project/resource/UserResource.java [hunk 3]
```diff
@@ -30,6 +33,17 @@ public UserResource(UserService userService) {
         return ResponseEntity.ok(user);
     }
 
+    @PutMapping("/user/add/friend")
+    public ResponseEntity<UserDto> addUserFriends(@RequestBody UserDto userDto) {
+        final SecurityContext context = SecurityContextHolder.getContext();
+        final Authentication authentication = context.getAuthentication();
+        final String username = authentication.getName();
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/resource/UserResource.java:40]: Move to service layer

File: src/main/java/com/kpi/project/validate/UserValidator.java [hunk 4]
```diff
@@ -52,6 +52,12 @@ public void validateUserExistence(Long id) {
         }
     }
 
+    public void validateUserFriend(String username) {
+        if (Objects.isNull(userRepository.findByUsername(username))) {
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/validate/UserValidator.java:56]: make it simpler with user==null

File: src/main/java/com/kpi/project/validate/UserValidator.java [hunk 5]
```diff
@@ -52,6 +52,12 @@ public void validateUserExistence(Long id) {
         }
     }
 
+    public void validateUserFriend(String username) {
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/validate/UserValidator.java:55]: validateFriendToAdd

File: src/main/java/com/kpi/project/model/dto/UserDto.java (hunk 2 above)
Comment (snahornyi) [reply src/main/java/com/kpi/project/model/dto/UserDto.java:24]: User can be returned with multiple friends

File: src/main/java/com/kpi/project/resource/UserResource.java [hunk 6]
```diff
@@ -30,6 +33,17 @@ public UserResource(UserService userService) {
         return ResponseEntity.ok(user);
     }
 
+    @PutMapping("/user/add/friend")
+    public ResponseEntity<UserDto> addUserFriends(@RequestBody UserDto userDto) {
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/resource/UserResource.java:37]: As discussed we need just friend username

File: src/main/java/com/kpi/project/service/UserService.java [hunk 7]
```diff
@@ -56,6 +57,20 @@ public UserDto updateUserRoles(UserDto userDto) throws UsernameNotFoundException
         return userMapper.userToDto(userRepository.save(userWithNewRoles));
     }
 
+    public UserDto addUserFriend(String username, String friendName) throws UsernameNotFoundException {
+        userValidator.validateUserFriend(friendName);
+
+        final Set<User> friends = userRepository.findByUsername(username).getFriends();
+        final Set<User> updatedFriends = friends != null ? friends : new HashSet<>();
+        updatedFriends.add(userRepository.findByUsername(friendName));
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/service/UserService.java:65]: Two calls of userRepository.findByUsername(friendName)
1) Here
2) Validator

File: src/test/java/com/kpi/project/validator/UserValidatorTest.java [hunk 8]
```diff
@@ -146,6 +146,17 @@ public void validateUserExistenceShouldThrowExceptionUserIsNotExist() {
...
 
+    @Test
+    public void validateUserFriendShouldThrowExceptionUserIsNotExist() {
+        //given
+        given(userRepository.findByUsername(any())).willReturn(null);
+
+        //expected
+        assertThatExceptionOfType(ValidatorException.class)
+                .isThrownBy(() -> testingInstance.validateUserFriend("username"))
+                .withMessage("User with username : username, not exists");
```
Comment (snahornyi) [inline src/test/java/com/kpi/project/validator/UserValidatorTest.java:157]: "User with username: username, not exists"

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/com/kpi/project/model/dto/UserDto.java [hunk 9]
```diff
@@ -21,4 +22,6 @@
 
     Set<String> roles;
 
+    Set<User> friendNames;
```
Comment (snahornyi) [inline src/main/java/com/kpi/project/model/dto/UserDto.java:25]: It's whole users, not just names

**Review by snahornyi — CHANGES_REQUESTED** [review]

**Review by snahornyi — APPROVED** [review]

**Review by Vlad6501 — APPROVED** [review]

**Pull request merged**